### PR TITLE
First draft for api usability

### PR DESCRIPTION
**Pull request by anthliu**

### CODE REVIEW

File: src/language-usability.rst [hunk 1]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+..
+  Outline
+  * APIs are ubiquitous
+  * Examples
+  * High level goals of API
+  * Difficulty in achieving goals of API
+  * Distinction with language usability
+..
+
+Application Programming Interfaces (APIs) are a set of functions and procedures
```
Comment (cyrus-) [inline src/language-usability.rst]: The "functions and procedures" definition from the dictionary is not a good one. Consider "are interfaces that mediate interactions between software components".

File: src/language-usability.rst [hunk 2]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+that enable an application to access services or data from another source.
+Examples include software libraries and frameworks, OS system calls, and web protocols.
+APIs are a ubiquitous aspect of programming.
```
Comment (cyrus-) [inline src/language-usability.rst]: You can drop this sentence, since you're saying the same thing in the next sentence.

File: src/language-usability.rst [hunk 3]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+Almost every program written will use an API :cite:`myers2016improving`.
```
Comment (cyrus-) [inline src/language-usability.rst]: "Almost" -> "Nearly"

File: src/language-usability.rst [hunk 4]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+
+The goal of the API designer is to create an API that a broad range of
+programmers can use for their applications.
```
Comment (cyrus-) [inline src/language-usability.rst]: "for their applications" -> "in their applications"

File: src/language-usability.rst [hunk 5]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+This goal mirrors the goals of programming language designers --
+an API designer must a create a service that is "usable" to
+a broad and vaguely scoped audience.
+Like languages, creating APIs is a challenging design problem of making
```
Comment (cyrus-) [inline src/language-usability.rst]: "Like languages, APIs must be simple enough..."

File: src/language-usability.rst [hunk 6]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+a complex service simple enough to be used
+by users who may be are unaware of the inner workings of the API.
```
Comment (cyrus-) [inline src/language-usability.rst]: "may be are" -> "are"

File: src/language-usability.rst [hunk 7]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+
+However, unlike language design, which may begin with a "bottom-up" approach,
```
Comment (cyrus-) [inline src/language-usability.rst]: You should reword this to say that API design can sometimes be "bottom-up", like language design, and in other cases involves putting a more usable layer atop a more complex existing API.

File: src/language-usability.rst [hunk 8]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+designing APIs is often a problem of taking an already large and complex
+system and simplifying it for a user.
+This adds another layer of complexity in designing APIs.
+To illustrate this diffulty:
+
+    "The enormous size of public APIs contributes to these difficulties; for example, the Java Platform, Standard Edition API Specification includes more than 4,000 classes with more than 35,000 different methods, and Microsoft’s .NET Framework includes more than 140,000 classes, methods, properties, and fields."
+    :cite:`myers2016improving`.
+
+
+API Design Aspects
```
Comment (cyrus-) [inline src/language-usability.rst]: lets call this API Design Considerations rather than Aspects

File: src/language-usability.rst [hunk 9]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+by increasing complexity decreases the learnability of the API.
+However, increasing the learnability of an API by lowering the complexity also lowers the power of the API.
+
+.. container:: bib-item
+
+  .. bibliography:: api-usability.bib
+    :filter: key == 'myers2016improving'
+
+  This paper provides an overview of the design aspects of API design
+  (Learnability, Productivity, Error-Prevention, Simplicity, Consistency, and Matching Mental Models)
```
Comment (cyrus-) [inline src/language-usability.rst:82]: Covering these with a sentence or two in the text of this section would be useful.

File: src/language-usability.rst [hunk 10]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+
+.. container:: bib-item
+
+  .. bibliography:: api-usability.bib
+    :filter: key == 'scheller2015automated'
+
+  This paper identifies key aspects in API usability, in an *API Concepts Framework*,
+  a structure where *concepts* represent possible user actions, leading
+  to measureable properties on usability. Using this framework, the paper introduces
+  an automated method for measuring API usability, which they test in a user study.
```
Comment (cyrus-) [inline src/language-usability.rst]: What was the result of the user study?

File: src/language-usability.rst [hunk 11]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+  evaluating APIs, mitigating problems in APIs, and tools for improving API usability.
+
+.. container:: bib-item
+
+  .. bibliography:: api-usability.bib
+    :filter: key == 'scheller2015automated'
+
+  This paper identifies key aspects in API usability, in an *API Concepts Framework*,
+  a structure where *concepts* represent possible user actions, leading
+  to measureable properties on usability. Using this framework, the paper introduces
```
Comment (cyrus-) [inline src/language-usability.rst:94]: "measurable properties related to usability"

File: src/language-usability.rst [hunk 12]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+
+..
+  Outline
+  * User study studying global usability of APIs very difficult, so many factors
+  * Most user studies in API examine a specific aspect of api design and test it
+  * Paper: Usability Implications of Requiring Parameters in Objects' Constructors
+  * Paper: The Implications of Method Placement on API Learnability
+..
+
+Testing how to best design APIs is immensely difficult, as the numerous API design
```
Comment (cyrus-) [inline src/language-usability.rst]: "Testing how to best design APIs" -> "Testing API designs"
remove "immensely"

File: src/language-usability.rst [hunk 13]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+examine a specific aspect of API design :cite:`stylos2007usability`
+:cite:`stylos2008implications`.
+
+.. container:: bib-item
+
+  .. bibliography:: api-usability.bib
+    :filter: key == 'stylos2007usability'
+
+  This paper examines the impact of requiring parameters in
+  objects' constructors on usability. The study in this paper showed that developers
```
Comment (cyrus-) [inline src/language-usability.rst]: -> "This paper examines the usability implications of class constructor parameters."

File: src/language-usability.rst [hunk 14]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+  * Paper: Usability Implications of Requiring Parameters in Objects' Constructors
+  * Paper: The Implications of Method Placement on API Learnability
+..
+
+Testing how to best design APIs is immensely difficult, as the numerous API design
+decisions all affect the overall usability interdependently
+:cite:`myers2016improving`. Testing the high-level and global design decisions
+is often intractable because of this. Most user studies in API usability
+examine a specific aspect of API design :cite:`stylos2007usability`
+:cite:`stylos2008implications`.
```
Comment (cyrus-) [inline src/language-usability.rst]: Can you add the factory pattern usability study as well.

File: src/language-usability.rst [hunk 15]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+..
+  Outline
+  * API usability can also be improved by creating better tools for understanding and searching API's -- documentation
+  * Another distinction with language design - documentation is part of API. Using API is using documentation
+  * Paper: Jadeite
+  * Paper: Improving API Documentation Usability with Knowledge Pushing
+  * Paper: Mica: A Web-Search Tool for Finding API Components and Examples
+..
+
+API usability can also be improved by creating better tools for understanding and searching through APIs -- improving API documentation
```
Comment (cyrus-) [inline src/language-usability.rst]: remove " -- improving API documentation" and the citations on this sentence

File: src/language-usability.rst [hunk 16]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+:cite:`stylos2009improving` :cite:`dekel2009improving` :cite:`stylos2006mica`.
+Documentation represents another distinction of API design from language design.
+The usage of APIs is so closely tied with using the API documentation that
```
Comment (cyrus-) [inline src/language-usability.rst:166]: good point!

File: src/language-usability.rst [hunk 17]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+the documentation can be considered part of the API.
+A good documentation is integral in making an inherently complex API more learnable.
```
Comment (cyrus-) [inline src/language-usability.rst]: "A good documentation" -> "Good documentation"

File: src/language-usability.rst [hunk 18]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+
+.. container:: bib-item
+
+  .. bibliography:: api-usability.bib
+    :filter: key == 'stylos2009improving'
+
+  This paper presents Jadeite, a tool for improving API documentation by
+  letting API users add "pretend" classes or methods in places where
```
Comment (cyrus-) [inline src/language-usability.rst]: It isn't totally clear what this means, can you elaborate a bit in this description?

File: src/language-usability.rst [hunk 19]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+  user. This paper shows this phenomenon occurs in real programming, and
+  shows how a function decorator can help alert users of such messages.
+
+.. container:: bib-item
+
+  .. bibliography:: api-usability.bib
+    :filter: key == 'stylos2006mica'
+
+  This paper presents Mica, a tool for helping API users search through APIs.
+  Mica is built on Google and analyzes the content of Google results to
```
Comment (cyrus-) [inline src/language-usability.rst]: remove "is built on Google and"

File: src/language-usability.rst [hunk 20]
```diff
@@ -16,6 +16,171 @@ Usability of Type Systems
...
+  give more relevant results for programmers.
```
Comment (cyrus-) [inline src/language-usability.rst]: how was this measured?

**Review by cyrus- — CHANGES_REQUESTED** [review]
Looks good overall! Please take a look at my comments on how to improve the clarity of your writing and some additional things to cover / write for your second draft.

File: src/language-usability.rst (hunk 20 above)
Comment (anthliu) [reply src/language-usability.rst]: They didn't seem to have a user study comparing Mica vs. plain google. But, they did briefly mention a formative study looking at how developers use google for learning APIs, which Mica addresses the problems they found. I added these details in the new draft

**Comment by anthliu** [comment]
Added changes for 2nd draft!

File: src/api-usability.bib [hunk 21]
```diff
@@ -0,0 +1,61 @@
+@article{myers2016improving,
+  title={Improving API usability},
+  author={Myers, Brad A and Stylos, Jeffrey},
+  journal={Communications of the ACM},
+  year={2016},
+}
+
+
+@article{scheller2015automated,
+  title={Automated measurement of API usability: The API concepts framework},
```
Comment (cyrus-) [inline src/api-usability.bib]: LaTeX will change the capitalization of things like "API", so you should put braces around that throughout the .bib file: `{API}`

File: src/api-usability.bib [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+  author={Scheller, Thomas and K{\"u}hn, Eva},
+  journal={Information and Software Technology},
+  year={2015},
+}
+
+
+@inproceedings{stylos2007usability,
+  title={Usability implications of requiring parameters in objects' constructors},
+  author={Stylos, Jeffrey and Clarke, Steven},
+  booktitle={Proceedings of the 29th international conference on Software Engineering},
```
Comment (cyrus-) [inline src/api-usability.bib]: Capitalize "International" and "Conference" (don't need braces outside of the title)

File: src/api-usability.bib [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+  year={2007},
+}
+
+@inproceedings{stylos2008implications,
+  title={The implications of method placement on API learnability},
+  author={Stylos, Jeffrey and Myers, Brad A},
+  booktitle={Proceedings of the 16th ACM SIGSOFT International Symposium on Foundations of software engineering},
```
Comment (cyrus-) [inline src/api-usability.bib]: capitalize "Software Engineering"

File: src/api-usability.bib [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+  title={Mica: A web-search tool for finding api components and examples},
+  author={Stylos, Jeffrey and Myers, Brad A},
+  booktitle={Visual Languages and Human-Centric Computing (VL/HCC'06)},
+  year={2006},
+}
+
+@inproceedings{ellis2007factory,
+  title={The factory pattern in API design: A usability evaluation},
+  author={Ellis, Brian and Stylos, Jeffrey and Myers, Brad},
+  booktitle={Proceedings of the 29th international conference on Software Engineering},
```
Comment (cyrus-) [inline src/api-usability.bib]: Capitalize "International Conference"

File: src/language-usability.rst [hunk 25]
```diff
@@ -16,6 +16,193 @@ Usability of Type Systems
...
+an API designer must a create a service that is "usable" to
+a broad and vaguely scoped audience.
+Like languages, APIs must be simple enough to be used
+by users who are unaware of the inner workings of the API.
+
+In many cases, both language and API design can be approached through a "bottom-up" manner -- designed with smaller components
+that can be combined to create complex behavior.
+However, in other cases, designing APIs involves involves putting a more usable layer atop a more complex existing API.
+This adds another layer of complexity in designing APIs.
+To illustrate this diffulty:
```
Comment (cyrus-) [inline src/language-usability.rst]: Change to "to illustrate this difficulty, Myers et al [1] write: "

File: src/language-usability.rst [hunk 26]
```diff
@@ -16,6 +16,193 @@ Usability of Type Systems
...
+..
+  Outline
+  * APIs are ubiquitous
+  * Examples
+  * High level goals of API
+  * Difficulty in achieving goals of API
+  * Distinction with language usability
+..
+
+Application Programming Interfaces (APIs) are interfaces that mediate interactions between software components.
```
Comment (cyrus-) [inline src/language-usability.rst]: Let's try "Application Programming Interfaces (APIs) are user interfaces that serve a dual role: they are simultaneously the user interfaces that programmers use to interact with software components and formal interfaces that govern interactions between different software components."

File: src/language-usability.rst [hunk 27]
```diff
@@ -16,6 +16,193 @@ Usability of Type Systems
...
+Examples include software libraries and frameworks, OS system calls, and web protocols.
```
Comment (cyrus-) [inline src/language-usability.rst]: * "OS system calls" -> "OS system call interfaces"

File: src/language-usability.rst [hunk 28]
```diff
@@ -16,6 +16,193 @@ Usability of Type Systems
...
+Nearly every program written will use an API :cite:`myers2016improving`.
+
+The goal of the API designer is to create an API that a broad range of
```
Comment (cyrus-) [inline src/language-usability.rst]: Some APIs are very domain specific while others are intentionally general, so it's not the case that every API aims to serve a broad range of programmers. See if you can make that distinction more clear here.

File: src/language-usability.rst [hunk 29]
```diff
@@ -16,6 +16,193 @@ Usability of Type Systems
...
+programmers can use in their applications.
+This goal mirrors the goals of programming language designers --
+an API designer must a create a service that is "usable" to
+a broad and vaguely scoped audience.
+Like languages, APIs must be simple enough to be used
+by users who are unaware of the inner workings of the API.
+
+In many cases, both language and API design can be approached through a "bottom-up" manner -- designed with smaller components
```
Comment (cyrus-) [inline src/language-usability.rst]: APIs differ from PLs in that PLs are axiomatic: a language definition defines a universe of programs. APIs operate within a programming language by making use of its primitive operations as well as other APIs.

File: src/language-usability.rst (hunk 25 above)
Comment (cyrus-) [reply src/language-usability.rst]: I think maybe this quote would be better in a new paragraph explaining that individual APIs are often organized into frameworks, some of which are very large:

File: src/language-usability.rst [hunk 30]
```diff
@@ -16,6 +16,193 @@ Usability of Type Systems
...
+
+There are several design aspects to desiging usable APIs.
+A broad aspect of usability identified by both :cite:`myers2016improving`
+and :cite:`scheller2015automated` is the effect of *complexity* on usability.
+Changing the inherant complexity exhibits one of the many tradeoffs between design aspects of APIs.
+Making an API more powerful (allowing a user to do more things with the API)
+by increasing complexity decreases the learnability of the API.
+However, increasing the learnability of an API by lowering the complexity also lowers the power of the API.
+
+Myers et al. identify 6 key design aspects of API design, which we describe here.
```
Comment (cyrus-) [inline src/language-usability.rst]: Introduce these things before the discussion on complexity above -- right now you talk about learnability and complexity (opposite of simplicity) before defining them.

File: src/language-usability.rst [hunk 31]
```diff
@@ -16,6 +16,193 @@ Usability of Type Systems
...
+
+* Learnability - an API user must learn how to use an API as quickly and easily as possible.
```
Comment (cyrus-) [inline src/language-usability.rst]: must -> should be able to learn (since you use should in the rest of these)

File: src/language-usability.rst [hunk 32]
```diff
@@ -16,6 +16,193 @@ Usability of Type Systems
...
+
+.. container:: bib-item
+
+  .. bibliography:: api-usability.bib
+    :filter: key == 'scheller2015automated'
+
+  This paper identifies key aspects in API usability, in an *API Concepts Framework*,
+  a structure where *concepts* represent possible user actions, leading
+  to measureable properties on usability. Using this framework, the paper introduces
+  an automated method for measuring API usability, which they test in a user study and
```
Comment (cyrus-) [inline src/language-usability.rst]: can you say a little more about this "automated method"? how does it work?

File: src/language-usability.rst [hunk 33]
```diff
@@ -16,6 +16,193 @@ Usability of Type Systems
...
+  * Most user studies in API examine a specific aspect of api design and test it
+  * Paper: Usability Implications of Requiring Parameters in Objects' Constructors
+  * Paper: The Implications of Method Placement on API Learnability
+..
+
+Testing API design is difficult, as the numerous API design
+decisions all affect the overall usability interdependently
+:cite:`myers2016improving`. Testing the high-level and global design decisions
+is often intractable because of this. Most user studies in API usability
+examine a specific aspect of API design :cite:`stylos2007usability`
```
Comment (cyrus-) [inline src/language-usability.rst]: Don't need these :cite:s here

File: src/language-usability.rst [hunk 34]
```diff
@@ -16,6 +16,193 @@ Usability of Type Systems
...
+:cite:`stylos2008implications` :cite:`ellis2007factory`.
+
+.. container:: bib-item
+
+  .. bibliography:: api-usability.bib
+    :filter: key == 'stylos2007usability'
+
+  This paper examines the usability implications of class constructor parameters.
```
Comment (cyrus-) [inline src/language-usability.rst]: add "in Java" to the end of the first sentence.

File: src/language-usability.rst [hunk 35]
```diff
@@ -16,6 +16,193 @@ Usability of Type Systems
...
+.. container:: bib-item
+
+  .. bibliography:: api-usability.bib
+    :filter: key == 'stylos2008implications'
+
+  This paper examines the impact of method placement on usability.
+  In the user study participants were given nearly identical APIs but
+  the comparison group contained a method that was placed in an
+  unexpected location. Participants with "correctly-placed" methods
+  were orders of magnitudes faster in their tasks.
```
Comment (cyrus-) [inline src/language-usability.rst]: "orders of magnitudes" is vague -- can you provide some numbers? what are "their tasks"?

File: src/language-usability.rst [hunk 36]
```diff
@@ -16,6 +16,193 @@ Usability of Type Systems
...
+.. container:: bib-item
+
+  .. bibliography:: api-usability.bib
+    :filter: key == 'stylos2009improving'
+
+  This paper presents Jadeite, a tool for improving API documentation.
+  In many APIs, users often expect a class or method to be placed in a
+  different location than actual, slowing them down :cite:`stylos2007usability`.
+  Jadeite allows API users to add "placeholder" classes or methods in places where
+  they expect a class or method to exist, which can redirect future API users to
```
Comment (cyrus-) [inline src/language-usability.rst]: remove "future"

**Review by cyrus- — CHANGES_REQUESTED** [review]
Looks good overall -- see comments for final draft. Grade will be sent by email.

**Comment by anthliu** [comment]
Thanks for the additional comments! I've added my changes for the final draft

**Review by cyrus- — APPROVED** [review]
great, thanks. will send final grades via email soon
